searx/infopage/__init__.py

```diff
@@ -0,0 +1,179 @@
+        .. _mistletoe: https://github.com/miyuchina/mistletoe
+
+        """
+        return mistletoe.markdown(self.content)
+
+    def get_ctx(self):  # pylint: disable=no-self-use
+        """Jinja context to render :py:obj:`InfoPage.content`"""
+
+        def _md_link(name, url):
+            url = url_for(url, _external=True)
+            return "[%s](%s)" % (name, url)
+
+        def _md_search(query):
+            url = '%s?q=%s' % (url_for('search', _external=True), urllib.parse.quote(query))
+            return '[%s](%s)' % (query, url)
+
+        ctx = {}
+        ctx['GIT_URL'] = GIT_URL
+        ctx['get_setting'] = get_setting
+        ctx['link'] = _md_link
+        ctx['search'] = _md_search
+
+        return ctx
+
+    def __repr__(self):
+        return f'<{self.__class__.__name__} fname={self.fname!r}>'
+
+
+class InfoPageSet:  # pylint: disable=too-few-public-methods
+    """Cached rendering of the online documentation a SearXNG instance has.
+
+    :param page_class: render online documentation by :py:obj:`InfoPage` parser.
+    :type page_class: :py:obj:`InfoPage`
+
+    :param info_folder: information directory
+    :type info_folder: str
+    """
+
+    def __init__(
+        self, page_class: typing.Optional[typing.Type[InfoPage]] = None, info_folder: typing.Optional[str] = None
+    ):
+        self.page_class = page_class or InfoPage
+        self.CACHE: typing.Dict[tuple, typing.Optional[InfoPage]] = {}
+
+        # future: could be set from settings.xml
+
+        self.folder: str = info_folder or _INFO_FOLDER
+        """location of the Markdwon files"""
+
+        self.locale_default: str = 'en'
+        """default language"""
+
+        self.locales: typing.List = [locale for locale in os.listdir(_INFO_FOLDER) if locale in LOCALE_NAMES]
+        """list of supported languages (aka locales)"""
+
+        self.toc: typing.List = [
+            'search-syntax',
+            'about',
+        ]
+        """list of articles in the online documentation"""
+
+    def get_page(self, pagename: str, locale: typing.Optional[str] = None):
+        """Return ``pagename`` instance of :py:obj:`InfoPage`
+
+        :param pagename: name of the page, a value from :py:obj:`InfoPageSet.toc`
+        :type pagename: str
+
+        :param locale: language of the page, e.g. ``en``, ``zh_Hans_CN``
+                       (default: :py:obj:`InfoPageSet.i18n_origin`)
+        :type locale: str
+
+        """
+        locale = locale or self.locale_default
+
+        if pagename not in self.toc:
+            return None
+        if locale not in self.locales:
+            return None
+
+        cache_key = (pagename, locale)
+        page = self.CACHE.get(cache_key)
+
+        if page is not None:
+            return page
+
+        # not yet instantiated
+
+        fname = os.path.join(self.folder, locale, pagename) + '.md'
+        if not os.path.exists(fname):
+            logger.info('file %s does not exists', fname)
```

Comment:
A translation that doesn't exist is not error: a someone can translate only one page.